lib/routes/recovery-keys.js

```diff
@@ -0,0 +1,230 @@
+            throw errors.unverifiedSession()
+          }
+
+          return db.getRecoveryKey(uid, recoveryKeyId)
+            .then((res) => recoveryData = res.recoveryData)
+        }
+      }
+    },
+    {
+      method: 'POST',
+      path: '/account/reset/recoveryKeys',
+      config: {
+        auth: {
+          strategy: 'accountResetToken'
+        },
+        validate: {
+          query: {
+            keys: isA.boolean().optional()
+          },
+          payload: {
+            authPW: isA.string().min(64).max(64).regex(HEX_STRING).required(),
+            wrapKb: isA.string().min(64).max(64).regex(HEX_STRING).required(),
+            recoveryKeyId: isA.string().max(256).required(),
+          }
+        }
+      },
+      handler: function (request, reply) {
+        log.begin('Password.changeRecoveryKeys', request)
+        const authPW = request.payload.authPW
+        const wrapKb = request.payload.wrapKb
+        const recoveryKeyId = request.payload.recoveryKeyId
+        const uid = request.auth.credentials.uid
+
+        let authSalt, password, verifyHash, sessionToken, keyFetchToken, account
+
+        checkRecoveryKey()
+          .then(getAccount)
+          .then(passwordReset)
+          .then(deleteRecoveryKey)
+          .then(notifyAccount)
+          .then(createSessionToken)
+          .then(createKeyFetchToken)
+          .then(createResponse)
+          .then(reply, reply)
+
+        function checkRecoveryKey() {
+          return db.getRecoveryKey(uid, recoveryKeyId)
+        }
+
+        function getAccount() {
+          return db.account(uid)
+            .then((result) => account = result)
+        }
+
+        function passwordReset() {
+          return random.hex(32)
+            .then(hex => {
+              authSalt = hex
+              password = new Password(authPW, authSalt, verifierVersion)
+              return password.verifyHash()
+            })
+            .then((hash) => {
+              verifyHash = hash
+              return password.wrap(wrapKb)
+            })
+            .then((wrapWrapKb) => {
+              return db.resetAccount(
+                {uid},
+                {
+                  verifyHash: verifyHash,
+                  authSalt: authSalt,
+                  wrapWrapKb: wrapWrapKb,
+                  verifierVersion: password.version
+                }
+              )
+            })
+        }
+
+        function deleteRecoveryKey() {
+          return db.deleteRecoveryKey(uid, recoveryKeyId)
+        }
+
+        function notifyAccount() {
+          return Promise.resolve()
+        }
+
+        function createSessionToken() {
+          const {
+            browser: uaBrowser,
+            browserVersion: uaBrowserVersion,
+            os: uaOS,
+            osVersion: uaOSVersion,
+            deviceType: uaDeviceType,
+            formFactor: uaFormFactor
+          } = request.app.ua
+
+          const sessionTokenOptions = {
```

Comment:
I see this same paradigm of getting a bunch of fields from request.app.ua and then using those to set up sessionTokenOptions in 4 other places in the tree, time to abstract that to minimize the chance of forgetting to add or remove one of those generic fields in the future?

Reply:
I took a look at refactoring this and I think it is big enough to be done in its own PR, https://github.com/mozilla/fxa-auth-server/issues/2481.